0000-prelude-2024-future.md

```diff
@@ -0,0 +1,94 @@
+- Feature Name:  `prelude_2024_future`
+- Start Date:  2023-10-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC describes the inclusion of the `Future` and `IntoFuture` traits in the 2024 edition prelude.
+
+# Motivation
+[motivation]: #motivation
+
+When an `async fn` is desugared we obtain an anonymous type with a signature of `impl Future`. In order to use this type we must first be able to _name_ it, which can only be done if the  `Future` trait in scope. Currently this can be a little inconvenient since it requires `std::future::Future` to be manually imported.
+
+Most other reifications of control-flow effects have their respective types and traits included in the prelude header. To support iteration we include both the `Iterator` and `IntoIterator` traits in the prelude. To support fallibility we include both `Option` and `Result` in the prelude. This RFC proposes we include both `Future` and `IntoFuture` to match.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Let's say someone wrote an async function which takes a future and operates on it. For example, it could be something like this:
+
+```rust
+use std::future::IntoFuture;
+
+async fn meow(phrase: impl IntoFuture<Output = String>) {
+    println!("{}, meow", phrase.await);
+}
+```
+
+In the Rust 2021 edition this code would require an explicit import of the `IntoFuture` trait. When migrating to the 2024 edition the import of the `IntoFuture` trait would be taken care of by the prelude, and the code could remove the explicit import:
+
+```rust
+async fn meow(phrase: impl IntoFuture<Output = String>) {
+    println!("{}, meow", phrase.await);
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes we include both `Future` and `IntoFuture` as part of the 2024 prelude. The prelude in `std` re-exports the prelude in `core`. So the only change we need to make is in the prelude in `core`, changing it to both include `Future` and `IntoFuture`:
+
+```rust
+// core::prelude::rust_2024
+mod rust_2024 {
+    pub use crate::future::Future;
+    pub use crate::future::IntoFuture;
+    pub use super::rust_2021::*; 
+}
+```
+
+# Tradeoffs
+[tradeoffs]: #tradeoffs
+
+Both the `Future` and `IntoFuture` definitions in the standard library are considered _canonical_: there exist no widespread alternative definitions in the Rust ecosystem. Simply having both traits in scope is unlikely to lead to any issues, and the only likely noticable outcome is that authoring async code will require slightly less effort.
+
+# Rationale and alternatives
```

Comment:
I think it'd be good to add a "Do nothing" section as an alternative. This is kind of addressed elsewhere in the RFC (limited expected breakage and async code is slightly harder to write), but having it called out into its own section would be helpful.